File: 27-Trie/cpp/04-number-of-distinct-substrings-in-a-string.cpp
#include <string>

using namespace std;

class Solution {
private:
    struct Node {
        Node* children[26] = {nullptr};
    };

public:
    int countDistinctSubstring(string s) {
        Node* root = new Node();
        int count = 0;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            Node* curr = root;
            for (int j = i; j < n; j++) {
                int idx = s[j] - 'a';
                if (!curr->children[idx]) {
                    curr->children[idx] = new Node();
                    count++;
                }
                curr = curr->children[idx];
            }
        }

        return count + 1;
    }
};

File: 27-Trie/java/04-NumberOfDistinctSubstringsInAString.java
class Solution {
    static class Node {
        Node[] children = new Node[26];
    }

    public int countDistinctSubstring(String s) {
        Node root = new Node();
        int count = 0;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            Node curr = root;
            for (int j = i; j < n; j++) {
                int idx = s.charAt(j) - 'a';
                if (curr.children[idx] == null) {
                    curr.children[idx] = new Node();
                    count++;
                }
                curr = curr.children[idx];
            }
        }

        return count + 1; // +1 for empty substring
    }
}
